src/server/services/records.js

```diff
@@ -126,8 +134,37 @@ async function recordsForUpload (upload) {
   return records
 }
 
+/**
+ * Wraps loadRecordsForUpload with per-request memoization. This ensures that
+ * the same upload is not read from the filesystem more than once during a
+ * single request.
+ *
+ * @param {object} upload The upload to fetch records for.
+ * @returns {Promise<object[]>} A list of records corresponding to the requested upload
+ */
+async function recordsForUpload (upload) {
+  log(`recordsForUpload(${upload.id})`)
+
+  const req = useRequest()
+  if (!req.recordsForUpload) {
+    req.recordsForUpload = {}
+  }
+  if (req.recordsForUpload[upload.id]) {
+    log(`recordsForUpload(${upload.id}): waiting for read`)
```

Comment:
nit: this log isn't quite accurate; we don't know whether the promise has resolved or not when we return it here

Reply:
Oh yeah, thanks!